In [1]:
!wget https://huggingface.co/datasets/Jgmorenof/teaching_tools_2025/resolve/main/chef-douvre.zip

--2026-01-26 20:40:48--  https://huggingface.co/datasets/Jgmorenof/teaching_tools_2025/resolve/main/chef-douvre.zip
Resolving huggingface.co (huggingface.co)... 3.170.185.33, 3.170.185.14, 3.170.185.25, ...
Connecting to huggingface.co (huggingface.co)|3.170.185.33|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://us.gcp.cdn.hf.co/xet-bridge-us/68a6cbb4eefddf148411cee1/54a51d564f259297a0705d4873bb2975d6589c79119ea8729eb93788a856ab4d?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27chef-douvre.zip%3B+filename%3D%22chef-douvre.zip%22%3B&response-content-type=application%2Fzip&Expires=1769463649&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiRXBvY2hUaW1lIjoxNzY5NDYzNjQ5fX0sIlJlc291cmNlIjoiaHR0cHM6Ly91cy5nY3AuY2RuLmhmLmNvL3hldC1icmlkZ2UtdXMvNjhhNmNiYjRlZWZkZGYxNDg0MTFjZWUxLzU0YTUxZDU2NGYyNTkyOTdhMDcwNWQ0ODczYmIyOTc1ZDY1ODljNzkxMTllYTg3MjllYjkzNzg4YTg1NmFiNGRcXD9yZXNwb25zZS1jb250ZW50LWRpc3Bvc2l0aW9uPSomcmVzcG9uc2UtY29udGV

In [2]:
!unzip -l chef-douvre.zip 'chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/*' | head -n 101 | tail -n 100 | xargs unzip chef-douvre.zip -d /content/

Archive:  chef-douvre.zip
   creating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18960115_1-0003.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19140115_1-0005.xml  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18980115_1-0001.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18840315_1-0002.xml  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18720715_1-0001.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18980115_1-0004.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18900115_1-0002.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19180115_1-0003.xml  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19140115_1-0005.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18740715_1-0004.xml  
  inflating: /content/chef-douvre/AS_TrainingSet_

In [3]:
import os
import glob

# 1. Spécifiez le chemin de votre dossier
# **⚠️ CHANGEZ CETTE LIGNE**
DOSSIER = '/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2'

# 2. Trouvez tous les fichiers JPG/JPEG (ignore la casse)
fichiers_jpg = glob.glob(os.path.join(DOSSIER, '*.[jJ][pP][gG]'))
fichiers_jpeg = glob.glob(os.path.join(DOSSIER, '*.[jJ][pP][eE][gG]'))

tous_les_fichiers = fichiers_jpg + fichiers_jpeg

# 3. Supprimez-les un par un
for fichier in tous_les_fichiers:
    try:
        os.remove(fichier)
        print(f"Supprimé : {fichier}")
    except OSError as e:
        print(f"Erreur lors de la suppression de {fichier} : {e}")

print(f"\nTerminé. {len(tous_les_fichiers)} fichier(s) ciblé(s).")

Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19020115_1-0006.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19240115_1-0004.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19060115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18680715_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18920115_1-0003.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18960115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18710715_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19000115_1-0004.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19280115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18980115_1-0004.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19040115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18680715_1-0004.jpg
Supprimé : /cont

In [4]:
import os
import json
import argparse
from pathlib import Path
from lxml import etree
from sentence_transformers import SentenceTransformer
import numpy as np
from tqdm import tqdm
from collections import defaultdict
from sklearn.metrics.pairwise import cosine_similarity

# ----------------- Constantes et Fonctions PAGE-XML -----------------
PAGE_NS = "http://schema.primaresearch.org/PAGE/gts/pagecontent/2013-07-15"
NSMAP = {"ns": PAGE_NS}

#def get_article_id_from_custom(custom_attr):
    #"""Extrait l'ID de l'article (e.g., 'a13') de l'attribut custom."""
"""
    if not custom_attr:
        return "N/A"
    # Recherche du motif 'structure {id:XXXX}'
    parts = custom_attr.split('structure {id:')
    if len(parts) > 1:
        # L'ID est après 'structure {id:' et avant la première '}'
        article_id = parts[1].split('}')[0]
        return article_id.strip()
    return "N/A" # Aucune ID d'article trouvée
"""

def get_article_id_from_custom(custom_attr):
    """Extrait l'ID de l'article (e.g., 'a13') de l'attribut custom."""
    if not custom_attr:
        return "N/A"
    # Recherche du motif 'structure {id:XXXX}'
    parts = custom_attr.split('structure {id:')
    if len(parts) > 1:
        # L'ID est après 'structure {id:' et avant le ';'
        article_id = parts[1].split(';')[0]
        return article_id.strip()
    return "N/A" # Aucune ID d'article trouvée

def coords_from_points(points_str):
    """
    points_str: "x1,y1 x2,y2 x3,y3 ..."
    returns xmin,ymin,xmax,ymax (floats)
    """
    pts = []
    for pair in points_str.strip().split():
        try:
            x_s,y_s = pair.split(",")
            pts.append((float(x_s), float(y_s)))
        except:
            continue
    xs = [p[0] for p in pts] if pts else [0.0]
    ys = [p[1] for p in pts] if pts else [0.0]
    return min(xs), min(ys), max(xs), max(ys)

def parse_page_xml(path):
    """
    Parse PAGE XML file et retourne la liste des blocs (un bloc = un TextRegion)
    avec l'ID d'article et le texte consolidé.
    [MODIFIÉ pour extraction de texte consolidé et article_id]
    """
    tree = etree.parse(str(path))
    root = tree.getroot()

    # Dimensions de la page
    page_elem = root.find(".//ns:Page", namespaces=NSMAP)
    image_w = float(page_elem.get("imageWidth")) if page_elem is not None and page_elem.get("imageWidth") else None
    image_h = float(page_elem.get("imageHeight")) if page_elem is not None and page_elem.get("imageHeight") else None
    blocks = []

    # Parcourir TextRegion
    for region in root.findall(".//ns:TextRegion", namespaces=NSMAP):

        # 1. Récupération de l'ID d'Article
        #custom_attr = region.get("custom")
        article_id_releve = False
        #Relever des textlines car l'id de l'article se trouve dans ces textlines
        textlines = region.findall(".//ns:TextLine", namespaces=NSMAP)
        tl = textlines[0]
        custom_attr = tl.get("custom")
        #print(custom_attr)
        article_id = get_article_id_from_custom(custom_attr)


        # 2. Récupération du Texte CONSOLIDÉ (Recherche du DERNIER TextEquiv/Unicode dans la Region)
        text = None
        # Chercher tous les TextEquiv/Unicode dans la région. Le dernier est souvent le texte final/consolidé.
        all_unicode_elements = region.findall(".//ns:TextEquiv/ns:Unicode", namespaces=NSMAP)

        if all_unicode_elements:
            # On prend le texte du dernier élément Unicode trouvé dans la région (correction de la troncature)
            text = all_unicode_elements[-1].text.strip() if all_unicode_elements[-1].text else None

        # 3. Fallback Texte: Concaténation des TextLine si aucun texte consolidé n'est trouvé
        if not text:
            #textlines = region.findall(".//ns:TextLine", namespaces=NSMAP)
            line_texts = []
            for tl in textlines:
                te = tl.find(".//ns:TextEquiv/ns:Unicode", namespaces=NSMAP)
                line_text = te.text.strip() if te is not None and te.text else "".join(tl.itertext()).strip()
                if line_text:
                    line_texts.append(line_text)

            text = " ".join(line_texts).strip()

        if not text:
            continue # Ignorer les régions sans texte

        # 4. Récupération des Coordonnées (Bounding Box de la Region)
        coords_node = region.find(".//ns:Coords", namespaces=NSMAP)
        if coords_node is not None and coords_node.get("points"):
            xmin, ymin, xmax, ymax = coords_from_points(coords_node.get("points"))
        else:
            xmin = ymin = 0.0; xmax = 1.0; ymax = 1.0

        # Normalisation
        if image_w and image_h:
            nxmin = xmin / image_w
            nxmax = xmax / image_w
            nymin = ymin / image_h
            nymax = ymax / image_h
        else:
            nxmin, nymin, nxmax, nymax = xmin, ymin, xmax, ymax

        blocks.append({
            "text": text,
            "article_id": article_id,
            "xmin": float(nxmin), "ymin": float(nymin),
            "xmax": float(nxmax), "ymax": float(nymax),
            "image_w": image_w, "image_h": image_h,
            "source": str(path),
            "id": region.get("id") # ID de la TextRegion
        })

    return blocks

In [5]:
# ----------------- Fonctions Utilitaires (Inchangées) -----------------

def save_load_embeddings(texts, embeddings_file_path, model_name='all-MiniLM-L6-v2', batch_size=64, device=None):
    """
    Tente de charger les embeddings à partir du chemin unique fourni.
    Si le fichier n'existe pas ou la taille ne correspond pas, recalcule et sauvegarde.
    """
    embeddings_file = Path(embeddings_file_path)
    non_empty_texts = [t for t in texts if t]

    if not non_empty_texts:
        print("[3] Aucun texte à intégrer (blocs vides).")
        return np.array([])

    try:
        embed_dim = SentenceTransformer(model_name).get_sentence_embedding_dimension()
    except Exception:
        embed_dim = 384
    expected_shape = (len(non_empty_texts), embed_dim)

    if embeddings_file.exists():
        print(f"[3] Chargement des embeddings depuis {embeddings_file}...")
        try:
            embeddings = np.load(embeddings_file)
            if embeddings.shape == expected_shape:
                print("    Chargement réussi et forme correspondante.")
                return embeddings
            print(f"    ATTENTION: Cache invalide. Recalcul...")
        except Exception as e:
            print(f"    Erreur lors du chargement/Vérification: {e}. Recalcul en cours...")

        try:
            os.remove(embeddings_file)
            print(f"    Fichier cache obsolète/corrompu supprimé: {embeddings_file}")
        except:
            pass

    print(f"[3] Calcul et sauvegarde des embeddings SBERT vers {embeddings_file}...")
    embeddings = compute_sbert_embeddings(non_empty_texts, model_name)

    try:
        np.save(embeddings_file, embeddings)
        print(f"    Embeddings sauvegardés avec succès.")
    except Exception as e:
        print(f"    AVERTISSEMENT: Échec de la sauvegarde des embeddings: {e}")

    return embeddings

def load_all_blocks(xml_folder):
    # ... (Votre fonction load_all_blocks inchangée) ...
    xml_folder = Path(xml_folder)
    all_blocks = []
    for f in xml_folder.glob("*.xml"):
        try:
            bls = parse_page_xml(f)
            all_blocks.extend(bls)
        except Exception as e:
            print(f"Erreur parsing {f}: {e}")
    return all_blocks

def merge_short_blocks(blocks, char_threshold=40, y_tol=0.03, x_tol=0.15):
    # ... (Votre fonction merge_short_blocks inchangée) ...
    """
    Blocks coords are normalized [0,1].
    Merge short blocks with a vertical neighbor in the same column.
    y_tol, x_tol are normalized tolerances.
    """
    blocks_by_source = defaultdict(list)
    for b in blocks:
        blocks_by_source[b["source"]].append(b)

    merged = []
    for src, bls in blocks_by_source.items():
        bls_sorted = sorted(bls, key=lambda b: (b["xmin"], b["ymin"]))
        used = [False]*len(bls_sorted)
        for i, b in enumerate(bls_sorted):
            if used[i]: continue

            text = b["text"]

            if len(text) < char_threshold:
                best_j = None
                for j in range(len(bls_sorted)):
                    if i == j or used[j]: continue

                    bj = bls_sorted[j]
                    x_overlap = not (bj["xmax"] < b["xmin"] or bj["xmin"] > b["xmax"])
                    same_col = x_overlap or abs(bj["xmin"] - b["xmin"]) < x_tol
                    vertically_close = abs(b["ymin"] - bj["ymax"]) < y_tol or abs(bj["ymin"] - b["ymax"]) < y_tol

                    if same_col and vertically_close:
                        best_j = j
                        break

                if best_j is not None:
                    bj = bls_sorted[best_j]
                    if b["ymin"] < bj["ymin"]:
                        new_text = b["text"] + " " + bj["text"]
                    else:
                        new_text = bj["text"] + " " + b["text"]

                    bj["text"] = new_text
                    bj["xmin"] = min(bj["xmin"], b["xmin"])
                    bj["ymin"] = min(bj["ymin"], b["ymin"])
                    bj["xmax"] = max(bj["xmax"], b["xmax"])
                    bj["ymax"] = max(bj["ymax"], b["ymax"])

                    used[i] = True
                else:
                    merged.append(b); used[i] = True
            else:
                merged.append(b); used[i] = True

    return merged

def compute_sbert_embeddings(texts, model_name='all-MiniLM-L6-v2', batch_size=64, device=None):
    model = SentenceTransformer(model_name, device=device)
    embeddings = model.encode(texts, batch_size=batch_size, show_progress_bar=True, convert_to_numpy=True)
    if embeddings.ndim != 2:
        raise ValueError(f"Embeddings have unexpected dimensions: {embeddings.ndim}")
    if embeddings.shape[0] != len(texts):
         raise ValueError(f"Number of embeddings ({embeddings.shape[0]}) does not match number of texts ({len(texts)})")
    return embeddings

In [6]:
merge_short = True
xml_folder = "/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2"
output_xml_folder = "/content/chef-douvre/output_sbert"
temp_data_folder = Path(xml_folder).parent / "temp_sbert_data"
os.makedirs(temp_data_folder, exist_ok=True)
os.makedirs(output_xml_folder, exist_ok=True)
char_threshold=40

"""
if clustering_kwargs is None:
    clustering_kwargs = {}
"""

print("[1] Extraction blocs depuis PAGE-XML...")
blocks = load_all_blocks(xml_folder)
# ... (Code de normalisation et de fusion) ...

if merge_short:
    print("[2] Fusion des petits blocs proches...")
    blocks_after_merge = merge_short_blocks(blocks, char_threshold=char_threshold)
else:
    blocks_after_merge = blocks

[1] Extraction blocs depuis PAGE-XML...
[2] Fusion des petits blocs proches...


In [7]:
len(blocks_after_merge)

11200

In [8]:
def creation_tab_a_partir_de_blocks(blocks):
  tab_texts = []
  tab_articles = []
  tab_coord_min = []
  tab_coord_max = []
  tab_sources = []
  dict_sum_sources_debut = {}
  dict_sum_sources_fin = {}
  tab_sources_diff = []
  sum = 0
  for k in range(len(blocks)):
    b = blocks[k]
    if k==0: #si on est au premier bloc
      dict_sum_sources_debut[b["source"]] = 0
    else:
      b2 = blocks[k-1]
      if (b2["source"]!=b["source"]):
        dict_sum_sources_debut[b["source"]] = sum
        dict_sum_sources_fin[b2["source"]] = sum-1 #car b2 précède b et sum est la valeur de la somme à b
      elif k==len(blocks)-1: #si on arrive au dernier bloc
        dict_sum_sources_fin[b["source"]]=sum
    sum += 1 #on ajoute 1 à la somme quand on passe au bloc suivant ou quand on s'arrête
  for b in blocks:
    tab_texts.append(b["text"])
    tab_articles.append(b["article_id"])
    tab_coord_min.append([b["xmin"],b["ymin"]])
    tab_coord_max.append([b["xmax"],b["ymax"]])
    tab_sources.append(b["source"])
    if b["source"] not in tab_sources_diff:
      tab_sources_diff.append(b["source"])
  return tab_texts, tab_articles, tab_coord_min, tab_coord_max, tab_sources, dict_sum_sources_debut, dict_sum_sources_fin, tab_sources_diff

In [9]:
tab_texts, tab_articles, tab_coord_min, tab_coord_max, tab_sources, dict_sum_sources_debut, dict_sum_sources_fin, tab_sources_diff = creation_tab_a_partir_de_blocks(blocks_after_merge)

In [10]:
len(tab_texts)

11200

In [11]:
len(tab_sources)

11200

**Embeddings**

multilingual-e5-large

In [12]:
import torch.nn.functional as F

from torch import Tensor
from transformers import AutoTokenizer, AutoModel


def average_pool(last_hidden_states: Tensor,
                 attention_mask: Tensor) -> Tensor:
    last_hidden = last_hidden_states.masked_fill(~attention_mask[..., None].bool(), 0.0)
    return last_hidden.sum(dim=1) / attention_mask.sum(dim=1)[..., None]


# Each input text should start with "query: " or "passage: ", even for non-English texts.
# For tasks other than retrieval, you can simply use the "query: " prefix.
#input_x = ["query: " + x]

tokenizer = AutoTokenizer.from_pretrained('intfloat/multilingual-e5-large-instruct')
model = AutoModel.from_pretrained('intfloat/multilingual-e5-large-instruct')

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json: 0.00B [00:00, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/690 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.12G [00:00<?, ?B/s]

In [14]:
import time

tab_embeddings = np.array([])
#tab_embeddings = []
i=0
#input = []
debut = time.time()
for text in tab_texts:
  #print("i = ",i)
  #input = ["query: " + text]
  input = ["passage: " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=512, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  """
  embeddings = embeddings.detach()
  tab_embeddings.append(embeddings)
  """

  embeddings = embeddings.detach().numpy()
  #tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)

  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
fin = time.time()
duree = fin - debut
duree_format_converti = time.strftime("%H:%M:%S", time.gmtime(duree))
print("Duree de creation des embeddings : ", duree_format_converti)
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Duree de creation des embeddings :  00:34:25


Durée : 35 minutes

In [15]:
tab_embeddings[0]

array([ 0.01669032,  0.0140363 , -0.02393063, ..., -0.04333923,
       -0.03796139,  0.01144409], dtype=float32)

In [16]:
len(tab_embeddings[0])

1024

In [17]:
output_path = Path(output_xml_folder)
os.makedirs(output_path, exist_ok=True)

import json

export_data = []
text_idx = 0
for i, block in enumerate(blocks_after_merge):
    text = block.get('text', '')
    if text:
        export_data.append({
            "index": text_idx,
            "text": text,
            "source_page": Path(block['source']).name,
            "article_id": block.get('article_id', 'N/A'),
            "embedding_vector": tab_embeddings[text_idx].tolist()
        })
        text_idx += 1

embeddings_json_file = output_path / "blocks_with_embeddings.json"
with open(embeddings_json_file, "w", encoding="utf-8") as f:
    json.dump(export_data, f, ensure_ascii=False, indent=2)
print(f"  => Données et embeddings exportés : {embeddings_json_file}")

  => Données et embeddings exportés : /content/chef-douvre/output_sbert/blocks_with_embeddings.json


In [ ]:
len(tab_embeddings)

In [ ]:
len(tab_texts)

**Similarité**

In [ ]:
from numpy.linalg import norm

def calcul_similarite(embedding1,embedding2): #calcul de la similarité entre 2 embeddings
  return np.dot(embedding1, embedding2) / (norm(embedding1)*norm(embedding2))

def tab_calcul_similarite(tab_embeddings): #calcul de la matrice des similarités
  if type(tab_embeddings)!=np.ndarray:
    tab_embeddings = np.array(tab_embeddings)
  #tab_similarites = np.zeros((len(tab_embeddings),len(tab_embeddings))) #matrice des similarités
  tab_similarites = []
  for k in range(len(tab_embeddings)): #calcul des similarités
    tab_sim = []
    for l in range(len(tab_embeddings)):
      shape_test1 = tab_embeddings[k].shape
      shape_test2 = tab_embeddings[l].shape
      embedding_test = tab_embeddings[l]
      if ((shape_test1[0]==shape_test2[0])&(tab_embeddings[l].ndim==2)):
        embedding_test = embedding_test.T
      sim = calcul_similarite(tab_embeddings[k],embedding_test)
      #tab_similarites[k][l] = sim
      tab_sim.append(sim)
    tab_similarites.append(tab_sim)
  tab_similarites = np.array(tab_similarites)
  return tab_similarites

def tab_calcul_similarite2(tab_embeddings,k,k_min,k_max):
  if (k_min > k):
    print("Erreur : k_min = ",k_min," > k = ",k)
    return
  if (k > k_max):
    print("Erreur : k = ",k," > k_max = ",k_max)
    return
  tab_emb = tab_embeddings[k_min:(k_max+1)]
  if type(tab_emb)!=np.ndarray:
    tab_emb = np.array(tab_emb)
  tab_similarites = np.zeros(len(tab_emb))
  for m in range(len(tab_emb)):
    if tab_emb[m].ndim==2:
      shape_test1 = tab_emb[k-k_min].shape
      shape_test2 = tab_emb[m].shape
      embedding_test = tab_emb[m]
      if (shape_test1[0]==shape_test2[0]):
        embedding_test = embedding_test.T
    tab_similarites[m] = calcul_similarite(tab_emb[k-k_min],tab_emb[m])
  return tab_similarites

**Vérifications**

In [ ]:
def recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=False,verif2=False): #fonction pour trouver l'embedding le plus similaire à un autre embedding
  if verif2:
    print("k = ",k)
  #pour garantir que l'ind_max donné vienne de la bonne source
  source = tab_sources[k]
  k_min = dict_sum_sources_debut[source]
  k_max = dict_sum_sources_fin[source]
  if verif2:
    print("k_min = ",k_min)
    print("k_max = ",k_max)
  tab_sim = tab_calcul_similarite2(tab_embeddings,k,k_min,k_max)
  tab_ind_potentiellement_valides = np.array([],dtype=int) #tableau des indices potentiellement valides selon les similarités calculées
  for m in range(k-k_min): #parcours pour les embeddings du même source avec indice entre k_min et k-1
    if tab_sim[m] >= seuil: #si la similarité entre ces 2 embeddings est supérieure au seuil donné, alors elle peut être prise en compte
      tab_ind_potentiellement_valides = np.append(tab_ind_potentiellement_valides,[m+k_min])
  for m in range(k-k_min+1,len(tab_sim)): #parcours pour les embeddings du même source avec indice entre k_min+1 et k-1
    if tab_sim[m] >= seuil:
      tab_ind_potentiellement_valides = np.append(tab_ind_potentiellement_valides,[m+k_min])
  for ind in tab_ind_potentiellement_valides:
    #calcul des distances entre les coordonnées
    dist_coord_min_x = abs(tab_coord_min[k][0] - tab_coord_min[ind][0])
    dist_coord_max_x = abs(tab_coord_max[k][0] - tab_coord_max[ind][0])
    dist_coord_min_y = abs(tab_coord_min[k][1] - tab_coord_min[ind][1])
    dist_coord_max_y = abs(tab_coord_max[k][1] - tab_coord_max[ind][1])
    dist_coord_x = [dist_coord_min_x,dist_coord_max_x]
    dist_coord_y = [dist_coord_min_y,dist_coord_max_y]
    #si les distances entre les coordonnées du bloc n°k et celles du bloc n°ind sont trop élevées
    #alors la similarité de leurs embeddings n'est pas prise en compte
    #if ((max(dist_coord_x)>0.0075)&(max(dist_coord_y)>0.05)):
    if ((max(dist_coord_x)>0.0025)&(max(dist_coord_y)>0.0125)): #j'ai mis ces valeurs pour augmenter la précision
                                                                #en minimisant le plus possible les différences entre les coordonnées
                                                                #l'inconvénient est que ça peut éliminer des exceptions valables
                                                                #mais pour ceux-là je peux utiliser le max de leurs similarités avec les autres embeddings du même source
      tab_ind_potentiellement_valides = np.delete(tab_ind_potentiellement_valides,np.where(tab_ind_potentiellement_valides == ind))
  if len(tab_ind_potentiellement_valides)>0: #si cette liste n'est pas vide
    #tab_sim2 = np.zeros(len(tab_ind_potentiellement_valides))
    ind_max = 0
    sim_max = 0
    for i in range(len(tab_ind_potentiellement_valides)): #prendre les similarités avec comme indice les éléments de la liste
      #id = tab_ind_potentiellement_valides[i] - k_min
      #tab_sim2[i] = tab_sim[id]
      id = tab_ind_potentiellement_valides[i]
      """
      if verif2:
        print("id = ",id)
      """
      if tab_sim[id-k_min] > sim_max:
        sim_max = tab_sim[id-k_min]
        ind_max = id
    #ind_max = tab_ind_potentiellement_valides[np.argmax(tab_sim2)] + k_min #prendre la similarité max
  else: #si cette liste est vide
    if verif: #si affichage possible
      print("Erreur pour embedding n°",k," : aucune similarite valide")
    tab_sim = np.delete(tab_sim,k-k_min) #car sinon on aura argmax(tab_sim) = k - k_min
    ind_max = np.argmax(tab_sim) + k_min #prendre la similarité max
    if ind_max >= k: #car tab_sim a pris en compte les embeddings n°k_min à k-1 et n°k+1 à k_max
      ind_max += 1
  if verif2:
    print("ind_max = ",ind_max)
    if ind_max > k_max:
      print("Erreur : ind_max > k_max")
      return
  return ind_max

In [ ]:
def verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,seuil,verif=False,verif2=False):
  nb_calcul_sim_valid = 0
  #tab_similarites = tab_calcul_similarite(tab_embeddings) #calcul de la matrice des similarités
  for k in range(len(tab_embeddings)): #pour chaque embedding
    if verif&verif2:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=verif,verif2=verif2)
    elif verif:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=verif)
    elif verif2:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif2=verif2)
    else:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil)
    if tab_articles[k] == tab_articles[ind_max]:
      nb_calcul_sim_valid += 1
  return nb_calcul_sim_valid / len(tab_embeddings)

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75)
precision

0.660625

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6635714285714286

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6608035714285714

*Avec max_length=512 et "query"*

if ((max(dist_coord_x)>0.0025)&(max(dist_coord_y)>0.0125)):

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6645535714285714

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6578571428571428

*Avec max_length=512 et "document"*

In [ ]:
tab_embeddings = np.array([])
i=0
for text in tab_texts:
  print("i = ",i)
  input = ["document : " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=512, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  embeddings = embeddings.detach().numpy()
  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Streaming output truncated to the last 5000 lines.
i =  6200
i =  6201
i =  6202
i =  6203
i =  6204
i =  6205
i =  6206
i =  6207
i =  6208
i =  6209
i =  6210
i =  6211
i =  6212
i =  6213
i =  6214
i =  6215
i =  6216
i =  6217
i =  6218
i =  6219
i =  6220
i =  6221
i =  6222
i =  6223
i =  6224
i =  6225
i =  6226
i =  6227
i =  6228
i =  6229
i =  6230
i =  6231
i =  6232
i =  6233
i =  6234
i =  6235
i =  6236
i =  6237
i =  6238
i =  6239
i =  6240
i =  6241
i =  6242
i =  6243
i =  6244
i =  6245
i =  6246
i =  6247
i =  6248
i =  6249
i =  6250
i =  6251
i =  6252
i =  6253
i =  6254
i =  6255
i =  6256
i =  6257
i =  6258
i =  6259
i =  6260
i =  6261
i =  6262
i =  6263
i =  6264
i =  6265
i =  6266
i =  6267
i =  6268
i =  6269
i =  6270
i =  6271
i =  6272
i =  6273
i =  6274
i =  6275
i =  6276
i =  6277
i =  6278
i =  6279
i =  6280
i =  6281
i =  6282
i =  6283
i =  6284
i =  6285
i =  6286
i =  6287
i =  6288
i =  6289
i =  6290
i =  6291
i =  6292
i =  6293
i =  6294

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Streaming output truncated to the last 5000 lines.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9956
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9953
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9933
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9588
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9983
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9958
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9977
k =  9962
k_min =  9522
k_max =  10043
ind_max =  10019
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
ind_max =  10010
k =  9967
k

0.6650892857142857

En comparaison, la précision obtenue quand on avait "query" à la place de "document" : 0.6578571428571428

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6683928571428571

En comparaison, la précision obtenue quand on avait "query" à la place de "document" : 0.6645535714285714

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6650892857142857

En comparaison, la précision obtenue quand on avait "query" à la place de "document" : 0.6578571428571428

*Avec max_length=256 et "document"*

In [ ]:
tab_embeddings = np.array([])
i=0
for text in tab_texts:
  print("i = ",i)
  input = ["document : " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=256, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  embeddings = embeddings.detach().numpy()
  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
i =  6200
i =  6201
i =  6202
i =  6203
i =  6204
i =  6205
i =  6206
i =  6207
i =  6208
i =  6209
i =  6210
i =  6211
i =  6212
i =  6213
i =  6214
i =  6215
i =  6216
i =  6217
i =  6218
i =  6219
i =  6220
i =  6221
i =  6222
i =  6223
i =  6224
i =  6225
i =  6226
i =  6227
i =  6228
i =  6229
i =  6230
i =  6231
i =  6232
i =  6233
i =  6234
i =  6235
i =  6236
i =  6237
i =  6238
i =  6239
i =  6240
i =  6241
i =  6242
i =  6243
i =  6244
i =  6245
i =  6246
i =  6247
i =  6248
i =  6249
i =  6250
i =  6251
i =  6252
i =  6253
i =  6254
i =  6255
i =  6256
i =  6257
i =  6258
i =  6259
i =  6260
i =  6261
i =  6262
i =  6263
i =  6264
i =  6265
i =  6266
i =  6267
i =  6268
i =  6269
i =  6270
i =  6271
i =  6272
i =  6273
i =  6274
i =  6275
i =  6276
i =  6277
i =  6278
i =  6279
i =  6280
i =  6281
i =  6282
i =  6283
i =  6284
i =  6285
i =  6286
i =  6287
i =  6288
i =  6289
i =  6290
i =  6291
i 

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9956
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9953
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9933
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9588
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9983
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9958
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9977
k =  9962
k_min =  9522
k_max =  10043
ind_max =  10019
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6650892857142857

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6683928571428571

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6650892857142857

*Avec max_length=128 et "document"*

In [ ]:
tab_embeddings = np.array([])
i=0
for text in tab_texts:
  print("i = ",i)
  input = ["document : " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=128, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  embeddings = embeddings.detach().numpy()
  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
i =  6200
i =  6201
i =  6202
i =  6203
i =  6204
i =  6205
i =  6206
i =  6207
i =  6208
i =  6209
i =  6210
i =  6211
i =  6212
i =  6213
i =  6214
i =  6215
i =  6216
i =  6217
i =  6218
i =  6219
i =  6220
i =  6221
i =  6222
i =  6223
i =  6224
i =  6225
i =  6226
i =  6227
i =  6228
i =  6229
i =  6230
i =  6231
i =  6232
i =  6233
i =  6234
i =  6235
i =  6236
i =  6237
i =  6238
i =  6239
i =  6240
i =  6241
i =  6242
i =  6243
i =  6244
i =  6245
i =  6246
i =  6247
i =  6248
i =  6249
i =  6250
i =  6251
i =  6252
i =  6253
i =  6254
i =  6255
i =  6256
i =  6257
i =  6258
i =  6259
i =  6260
i =  6261
i =  6262
i =  6263
i =  6264
i =  6265
i =  6266
i =  6267
i =  6268
i =  6269
i =  6270
i =  6271
i =  6272
i =  6273
i =  6274
i =  6275
i =  6276
i =  6277
i =  6278
i =  6279
i =  6280
i =  6281
i =  6282
i =  6283
i =  6284
i =  6285
i =  6286
i =  6287
i =  6288
i =  6289
i =  6290
i =  6291
i 

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9956
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9953
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9933
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9588
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9983
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9958
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9977
k =  9962
k_min =  9522
k_max =  10043
ind_max =  10019
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6645535714285714

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6676785714285715

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6645535714285714

*Avec max_length=64 et "document*

In [ ]:
tab_embeddings = np.array([])
i=0
for text in tab_texts:
  print("i = ",i)
  input = ["document : " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=64, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  embeddings = embeddings.detach().numpy()
  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
i =  6200
i =  6201
i =  6202
i =  6203
i =  6204
i =  6205
i =  6206
i =  6207
i =  6208
i =  6209
i =  6210
i =  6211
i =  6212
i =  6213
i =  6214
i =  6215
i =  6216
i =  6217
i =  6218
i =  6219
i =  6220
i =  6221
i =  6222
i =  6223
i =  6224
i =  6225
i =  6226
i =  6227
i =  6228
i =  6229
i =  6230
i =  6231
i =  6232
i =  6233
i =  6234
i =  6235
i =  6236
i =  6237
i =  6238
i =  6239
i =  6240
i =  6241
i =  6242
i =  6243
i =  6244
i =  6245
i =  6246
i =  6247
i =  6248
i =  6249
i =  6250
i =  6251
i =  6252
i =  6253
i =  6254
i =  6255
i =  6256
i =  6257
i =  6258
i =  6259
i =  6260
i =  6261
i =  6262
i =  6263
i =  6264
i =  6265
i =  6266
i =  6267
i =  6268
i =  6269
i =  6270
i =  6271
i =  6272
i =  6273
i =  6274
i =  6275
i =  6276
i =  6277
i =  6278
i =  6279
i =  6280
i =  6281
i =  6282
i =  6283
i =  6284
i =  6285
i =  6286
i =  6287
i =  6288
i =  6289
i =  6290
i =  6291
i 

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9956
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9953
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9933
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9588
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9983
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9958
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9962
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9977
k =  9962
k_min =  9522
k_max =  10043
ind_max =  10019
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6589285714285714

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6617857142857143

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6589285714285714

*Avec max_length=32 et "document*

In [ ]:
tab_embeddings = np.array([])
i=0
for text in tab_texts:
  print("i = ",i)
  input = ["document : " + text]
  # Tokenize the input texts
  batch_dict = tokenizer(input, max_length=32, padding=True, truncation=True, return_tensors='pt')
  outputs = model(**batch_dict)
  embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
  # normalize embeddings
  embeddings = F.normalize(embeddings, p=2, dim=1)
  embeddings = embeddings.detach().numpy()
  if i==0:
    tab_embeddings = embeddings
  else:
    tab_embeddings = np.append(tab_embeddings, embeddings, axis=0)
  i += 1
#precision_multilingual_e5_large = verif_similarites(tab_texts, tab_articles,tab_embeddings_multilingual_e5_large)
#precision_multilingual_e5_large

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
i =  6200
i =  6201
i =  6202
i =  6203
i =  6204
i =  6205
i =  6206
i =  6207
i =  6208
i =  6209
i =  6210
i =  6211
i =  6212
i =  6213
i =  6214
i =  6215
i =  6216
i =  6217
i =  6218
i =  6219
i =  6220
i =  6221
i =  6222
i =  6223
i =  6224
i =  6225
i =  6226
i =  6227
i =  6228
i =  6229
i =  6230
i =  6231
i =  6232
i =  6233
i =  6234
i =  6235
i =  6236
i =  6237
i =  6238
i =  6239
i =  6240
i =  6241
i =  6242
i =  6243
i =  6244
i =  6245
i =  6246
i =  6247
i =  6248
i =  6249
i =  6250
i =  6251
i =  6252
i =  6253
i =  6254
i =  6255
i =  6256
i =  6257
i =  6258
i =  6259
i =  6260
i =  6261
i =  6262
i =  6263
i =  6264
i =  6265
i =  6266
i =  6267
i =  6268
i =  6269
i =  6270
i =  6271
i =  6272
i =  6273
i =  6274
i =  6275
i =  6276
i =  6277
i =  6278
i =  6279
i =  6280
i =  6281
i =  6282
i =  6283
i =  6284
i =  6285
i =  6286
i =  6287
i =  6288
i =  6289
i =  6290
i =  6291
i 

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9956
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9966
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9968
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9971
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9538
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9960
k =  9958
k_min =  9522
k_max =  10043
ind_max =  9989
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9932
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9947
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9977
k =  9962
k_min =  9522
k_max =  10043
ind_max =  10011
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9976
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6351785714285715

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6360714285714286

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6351785714285715

Avec une petite modif de la création des blocs pour mieux distinguer les articles

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3